staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go

```diff
@@ -33,35 +34,121 @@ func NewCRDConverter(crd *apiextensions.CustomResourceDefinition) (safe, unsafe
+	// validVersions is list of valid versions for this CRD. both in and out objects will be validated toward this list first.
+	validVersions map[schema.GroupVersion]bool
+	// converter is the actual converter that converts an object in place to the given version.
+	converter crdConverterInterface
+	// safeConverter is the flag to set this converter to safe or unsafe one. If true, no input to the converter would be mutated.
+	safeConverter bool
+}
+
+var _ runtime.ObjectConvertor = &crdConverter{}
 
-// ConvertFieldLabel delegate the call to the unsafe converter.
-func (c *safeConverterWrapper) ConvertFieldLabel(version, kind, label, value string) (string, string, error) {
-	return c.unsafe.ConvertFieldLabel(version, kind, label, value)
+func (c *crdConverter) ConvertFieldLabel(version, kind, label, value string) (string, string, error) {
+	// We currently only support metadata.namespace and metadata.name.
+	switch {
+	case label == "metadata.name":
+		return label, value, nil
+	case !c.clusterScoped && label == "metadata.namespace":
+		return label, value, nil
+	default:
+		return "", "", fmt.Errorf("field label not supported: %s", label)
+	}
 }
 
-// Convert makes a copy of in object and then delegate the call to the unsafe converter.
-func (c *safeConverterWrapper) Convert(in, out, context interface{}) error {
-	inObject, ok := in.(runtime.Object)
+func (c *crdConverter) Convert(in, out, context interface{}) error {
+	unstructIn, ok := in.(runtime.Unstructured)
 	if !ok {
-		return fmt.Errorf("input type %T in not valid for object conversion", in)
+		return fmt.Errorf("input type %T in not valid for unstructured conversion", in)
 	}
-	return c.unsafe.Convert(inObject.DeepCopyObject(), out, context)
+
+	unstructOut, ok := out.(runtime.Unstructured)
+	if !ok {
+		return fmt.Errorf("output type %T in not valid for unstructured conversion", out)
+	}
+
+	outGVK := unstructOut.GetObjectKind().GroupVersionKind()
+	if !c.validVersions[outGVK.GroupVersion()] {
+		return fmt.Errorf("request to convert CRD from an invalid group/version: %s", outGVK.String())
+	}
+	inGVK := unstructIn.GetObjectKind().GroupVersionKind()
+	if !c.validVersions[inGVK.GroupVersion()] {
+		return fmt.Errorf("request to convert CRD to an invalid group/version: %s", inGVK.String())
+	}
+
+	unstructOut.SetUnstructuredContent(unstructIn.UnstructuredContent())
+	_, err := c.ConvertToVersion(unstructOut, outGVK.GroupVersion())
+	return err
 }
 
-// ConvertToVersion makes a copy of in object and then delegate the call to the unsafe converter.
-func (c *safeConverterWrapper) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
-	return c.unsafe.ConvertToVersion(in.DeepCopyObject(), target)
+// ConvertToVersion converts in object to the given gvk in place and returns the same `in` object.
+func (c *crdConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
+	var err error
+	if c.safeConverter {
```

Comment:
don't like to mix this into the crdConverter because being safe is orthogonal and can therefore be done compositionally.